### PR TITLE
short and long u2f names 

### LINKED ISSUE #4727: Can't register a U2F key with a long name
**Description**

It looks like there is a max number of characters for a U2F key depending on the environment. If I try to add a key with a 28 char. name in e.g. Chromium (for example: chaiyeb9fohngaGhapup2Xoh2Shu), the key is not triggered for confirmation. No error message is shown and the process doesn't fail. The client is stuck at waiting for the key. 

 - Browser: Chrome, and Chromium based browsers, and the Desktop client.
 - Version  v3.102.3 but possibly older versions as well
 - All operating systems

fails with more than:
 - 28 chars on Ubuntu Desktop
 - 20 chars on Ubuntu Chromium/ Chrome
 - 27 chars on Windows Desktop
 - 27 chars on Windows Chrome

**Note**
If you register a key with a name that would not work in chrome in firefox and then attempt to log in with it in chromium,
the login will fail with a "invalid credentials" error.

## Test Notes:
* [ ] test the following on:  [Ubuntu Firefox], [Ubuntu Chrome], [Windows Desktop Client], [Windows Chrome]
* [ ] go to settings, login
* [ ] click on register second factor
* [ ] select TOTP and check that "TOTP" is the default name
* [ ] select U2F and check that "U2F" is the default name
* [ ] try to register both second factors with a whitespace only name -> should use the default name
* [ ] for U2F: type in a name longer than 64 chars -> should display that the entered name is too long and clicking save will fail with an error dialog
* [ ] for TOTP: registering a second factor with a name longer than 64 chars works

### Regressions
* [ ] any newly registered second factors work for login
* [ ] any old second factors can still be used (use old desktop client to create some)

### LINKED ISSUE #4740: Can't register a U2F key without name
When setting up a two-factor authentication with a U2F key, it says that the name is optional. But the registration of the U2F key will not work without a name on Ubuntu Desktop.

**To Reproduce**
1. Use Ubuntu Desktop
2. Go to settings
3. Go to Login
4. Add a second factor
5. Select type "Security key (U2F)"
6. leave the name field empty and click save
7. the key is not triggered for confirmation

**Expected behavior**
It should be possible to register the U2F key without name or the text should be changed

**Desktop**
Ubuntu 20.04.5 LTS 

## Test Notes
see #4727

### PR DESCRIPTION
**Pull request by ganthern**
close #4727 
close #4740

### CODE REVIEW

**Comment by jomapp** [comment]
I see that this is closing #4789. Probably this is a mistake, right?

File: src/settings/login/secondfactor/SecondFactorEditDialog.ts [hunk 1]
```diff
@@ -124,128 +62,83 @@ export class SecondFactorEditDialog {
...
-				this.u2fRegistrationData = await this.webauthnClient.register(this.user._id, this.name)
-				this.verificationStatus = VerificationStatus.Success
-			} catch (e) {
-				console.log("Webauthn registration failed: ", e)
-				this.u2fRegistrationData = null
-				this.verificationStatus = VerificationStatus.Failed
-			}
+	private static showRecoveryInfoDialog(user: User) {
+		// We only show the recovery code if it is for the current user and it is a global admin
+		if (!isSameId(getEtId(logins.getUserController().user), getEtId(user)) || !user.memberships.find(gm => gm.groupType === GroupType.Admin)) {
```
Comment (charlag) [inline src/settings/login/secondfactor/SecondFactorEditDialog.ts:71]: this one sounds like it should go into the model

File: src/settings/login/secondfactor/SecondFactorEditModel.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+const DEFAULT_U2F_NAME = "U2F"
+const DEFAULT_TOTP_NAME = "TOTP"
+
+export enum NameValidationStatus {
+	Valid,
+	Invalid
+}
+
+/** Enum with user-visible types because we don't allow adding SecondFactorType.u2f anymore */
+export const FactorTypes = {
```
Comment (charlag) [inline src/settings/login/secondfactor/SecondFactorEditModel.ts:32]: maybe we should convert it to a true enum

File: src/settings/login/secondfactor/SecondFactorEditModel.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+	private u2fRegistrationData: U2fRegisteredDevice | null = null
+
+	constructor(
+		private readonly entityClient: EntityClient,
+		private readonly user: User,
+		private readonly mailAddress: string,
+		private readonly webauthnClient: WebauthnClient,
+		readonly totpKeys: TotpSecret,
+		private readonly webauthnSupported: boolean,
+		private readonly updateViewCallback: () => void = noOp
```
Comment (charlag) [inline src/settings/login/secondfactor/SecondFactorEditModel.ts:60]: why do we have default?

File: src/settings/login/secondfactor/SecondFactorEditModel.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	validationMessage(): TranslationKey | null {
+		return this.nameValidationStatus === NameValidationStatus.Valid
+			? null
+			: "textTooLong_msg"
+	}
+
+	getFactorTypesOptions() {
+		const options: Array<SelectorItem<FactorTypesEnum>> = []
```
Comment (charlag) [inline src/settings/login/secondfactor/SecondFactorEditModel.ts:113]: I don't think selector items belong to the model, I'd give a list of types

File: src/settings/login/secondfactor/SecondFactorEditModel.ts [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+		} else {
+			this.verificationStatus = VerificationStatus.Progress
+		}
+	}
+
+	/**
+	 * validates the input and makes the server calls to actually create a second factor
+	 * if the validation passes.
+	 */
+	async save(callback: (user: User) => void): Promise<void> {
```
Comment (charlag) [inline src/settings/login/secondfactor/SecondFactorEditModel.ts:172]: can we just return the user?

File: src/settings/login/secondfactor/SecondFactorEditModel.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+			_ownerGroup: this.user._ownerGroup,
+			name: this.name,
+		})
+
+		if (this.selectedType === FactorTypes.WEBAUTHN) {
+			sf.type = SecondFactorType.webauthn
+
+			if (this.verificationStatus !== VerificationStatus.Success) {
+				// noinspection ES6MissingAwait
+				Dialog.message("unrecognizedU2fDevice_msg")
```
Comment (charlag) [inline src/settings/login/secondfactor/SecondFactorEditModel.ts:202]: dialogs from models is not not great but if it's too hard to move around we can keep it. I'd just throw `UserError` and catch it on the client

File: src/settings/login/secondfactor/SecondFactorEditDialog.ts [hunk 7]
```diff
@@ -44,17 +44,23 @@ export class SecondFactorEditDialog {
...
 				view: () => this.render(),
 			},
-			okAction: () => this.model.save((user) => {
-				this.dialog.close()
-				SecondFactorEditDialog.showRecoveryInfoDialog(user)
-			}),
+			okAction: () => showProgressDialog(
+				"pleaseWait_msg",
+				// make sure any error messages displayed by save are shown on top of the progress dialog
+				delay(1).then(() => this.model.save((u) => this.finalize(u)))
```
Comment (charlag) [inline src/settings/login/secondfactor/SecondFactorEditDialog.ts:50]: we do we want error dialogs on top of progress dialogs? Don't we want it the other way around?

File: test/tests/settings/login/secondfactor/SecondFactorEditModelTest.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+	let entityClientMock: EntityClient
+	let userMock: User
+	let webAuthnClientMock: WebauthnClient
+	let loginFacadeMock: LoginFacade
+	const totpKeys = createTotpKeys()
+	const validName = "myU2Fkey"
+	const langMock: LanguageViewModel = object()
+	when(langMock.get(matchers.anything())).thenReturn("hello there")
+	// this is too long if you convert it to bytes
+	const invalidName = "🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏴‍☠️🏴‍☠️🏴‍☠️🏴‍☠️🏴‍☠️"
```
Comment (charlag) [inline test/tests/settings/login/secondfactor/SecondFactorEditModelTest.ts:34]: i object, I think it's a very cool name and I want to name my authenticators like that now

File: src/settings/login/secondfactor/SecondFactorEditModel.ts (hunk 3 above)
Comment (ganthern) [reply src/settings/login/secondfactor/SecondFactorEditModel.ts]: no idea

**Pull request merged**